### PR TITLE
Feat/migrate docs

### PR DESCRIPTION
**Pull request by 0xFable**
Adds documentation and examples of adding the `migrate` function to Cosmwasm Smart contracts. 
Also includes a lil bit of platform specific info to clear up confusion on terms. 

Can expanded or change up anything as needed 🤝

Used a number of cosmwasm contracts, the docs where I could and also this article [by Ethan](https://medium.com/cosmwasm/cosmwasm-for-ctos-ii-advanced-usage-ee04ce95d1d0) as sources and tried to collect one single doc with lots of examples and explainers.

Still some things needed on executing the migration I think. 

Related to https://github.com/InterWasm/DAO/issues/41
Possibly resolves InterWasm/DAO#41

### CODE REVIEW

File: docs/04-smart-contracts/03-migration/01-migration.md [hunk 1]
```diff
@@ -3,3 +3,212 @@ sidebar_position: 1
 ---
 
 # Migration
+
+Migration is the process through which a given smart contracts code can be swapped out or 'upgraded'.
+
+CosmWasm made contract migration a first-class experience. When instantiating a contract, there is an optional admin field that you can set. If it is left empty, the contract is immutable. If it is set (to an external account or governance contract), that account can trigger a migration. The admin can also change admin or even make the contract fully immutable after some time. However, when we wish to migrate from contract A to contract B, contract B needs to be aware somehow of how the state was encoded.
+
+This is where CW2 comes in. It specifies one special Singleton to be stored on disk by all contracts on instantiate. When the migrate function is called, then the new contract can read that data and see if this is an expected contract we can migrate from. And also contain extra version information if we support multiple migrate paths.
```
Comment (orkunkl) [inline docs/04-smart-contracts/03-migration/01-migration.md]:
```suggestion
This is where CW2 specification comes in. It specifies one special `Singleton` to be stored on disk by all contracts on instantiate. When the migrate function is called, then the new contract can read that data and see if this is an expected contract we can migrate from. And also contain extra version information if we support multiple migrate paths.
```

File: docs/04-smart-contracts/03-migration/01-migration.md [hunk 2]
```diff
@@ -3,3 +3,212 @@ sidebar_position: 1
...
+
+Working with CW2 is quite straightforward in that, as a smart contract developer you need only perform a couple of steps.
+
+The CW2 Spec provides a `set_contract_version` which should be used in instantiate to store the original version of a contract. It is important to also set_contract_version again after a successful migration to update it.
```
Comment (orkunkl) [inline docs/04-smart-contracts/03-migration/01-migration.md:15]: better to give code example inside instantiation here

Comment (0xFable) [reply docs/04-smart-contracts/03-migration/01-migration.md:15]: Added a code example with a minimal instantiate just to show setting.

**Comment by orkunkl** [comment]
Merging the PR, I will do a refactor later

**Pull request merged**